### PR TITLE
[Node-Agent] Ensure that local CA certificate directory has files before start copying.

### LINKED ISSUE #10171: `pull-gardener-e2e-kind`: Node may not get cordoned
**How to categorize this issue?**

/area testing
/kind bug

**What happened**:
Pattern of a failing build:
```
... condition type EveryNodeReady is not true yet, had message too many worker nodes are registered. Exceeding maximum desired machine count (6/3) with reason NodesScalingDown"}
... condition type EveryNodeReady is not true yet, had message too many worker nodes are registered. Exceeding maximum desired machine count (6/3) with reason NodesScalingDown"}
... condition type EveryNodeReady is not true yet, had message too many worker nodes are registered. Exceeding maximum desired machine count (6/3) with reason NodesScalingDown"}
... condition type EveryNodeReady is not true yet, had message too many worker nodes are registered. Exceeding maximum desired machine count (6/3) with reason NodesScalingDown"}
... condition type EveryNodeReady is not true yet, had message too many worker nodes are registered. Exceeding maximum desired machine count (4/3) with reason NodesScalingDown"}
... condition type EveryNodeReady is not true yet, had message too many worker nodes are registered. Exceeding maximum desired machine count (4/3) with reason NodesScalingDown"}
... condition type EveryNodeReady is not true yet, had message too many worker nodes are registered. Exceeding maximum desired machine count (4/3) with reason NodesScalingDown"}
... condition type EveryNodeReady is not true yet, had message too many worker nodes are registered. Exceeding maximum desired machine count (4/3) with reason NodesScalingDown"}
........
retry failed with context deadline exceeded, last error: shoot "e2e-default" was not successfully reconciled
```

Pattern of passing builds:
```
... condition type EveryNodeReady is not true yet, had message too many worker nodes are registered. Exceeding maximum desired machine count (6/3) with reason NodesScalingDown"}
... condition type EveryNodeReady is not true yet, had message 3 machines are waiting to be completely drained from pods. If this persists, check your pod disruption budgets and pending finalizers. Please note, that nodes that fail to be drained will be deleted automatically with reason NodesScalingDown"}
... condition type EveryNodeReady is not true yet, had message 2 machines are waiting to be completely drained from pods. If this persists, check your pod disruption budgets and pending finalizers. Please note, that nodes that fail to be drained will be deleted automatically with reason NodesScalingDown"}
... condition type EveryNodeReady is not true yet, had message 1 machine is waiting to be completely drained from pods. If this persists, check your pod disruption budgets and pending finalizers. Please note, that nodes that fail to be drained will be deleted automatically with reason NodesScalingDown"}
... condition type EveryNodeReady is not true yet, had message 1 machine is waiting to be completely drained from pods. If this persists, check your pod disruption budgets and pending finalizers. Please note, that nodes that fail to be drained will be deleted automatically with reason NodesScalingDown"}
```

This can be explained by existence of  a node that does not get cordoned when upgrading a worker pool. 

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:

- Gardener version:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- Others:

### LINKED ISSUE #10227: Node with `critical-components-not-ready` pod will be reported as `NodesScalingDown` in `EveryNodeReady` condition
**How to categorize this issue?**

/area quality
/kind bug

**What happened**:
While trying to fix https://github.com/gardener/gardener/issues/10171 the message
```
condition type EveryNodeReady is not true yet, had message too many worker nodes are registered. Exceeding maximum desired machine count (4/3) with reason NodesScalingDown
```
was making me assume that there was a problem with the scaling down logic. But after deeper digging it was an error in a node-critical `apiserver-proxy` pod.

**What you expected to happen**:
The condition to be false with reason `NodesScalingUp`

**How to reproduce it (as minimally and precisely as possible)**:
1. Create a local shoot cluster with 1 machine
2. Deploy this to the shoot:
```
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: busybox-daemonset
  namespace: kube-system
spec:
  selector:
    matchLabels:
      name: busybox
  template:
    metadata:
      labels:
        name: busybox
        node.gardener.cloud/critical-component: "true"
    spec:
      tolerations:
      - key: node.gardener.cloud/critical-components-not-ready
        operator: Exists
        effect: NoSchedule
      containers:
      - name: busybox
        image: busybox
        command: [ "/bin/sh", "-c", "--" ]
        args: [ "exit 1" ]
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: kubernetes.io/hostname
                operator: NotIn
                values:
                - <shoot_node_name>
```

3. Update `.spec.kubernetes.version`
4. State of machines and nodes:
```
$ k get machine -A
NAMESPACE             NAME                                    STATUS    AGE     NODE
shoot--local--local   shoot--local--local-local-78c8b-lh2cc   Pending   5m34s   machine-shoot--local--local-local-78c8b-lh2cc
shoot--local--local   shoot--local--local-local-79987-w6bjj   Running   20m     machine-shoot--local--local-local-79987-w6bjj

$ k get nodes
NAME                                            STATUS   ROLES    AGE     VERSION
machine-shoot--local--local-local-78c8b-lh2cc   Ready    <none>   5m57s   v1.30.0
machine-shoot--local--local-local-79987-w6bjj   Ready    <none>   21m     v1.29.0

$ k get nodes machine-shoot--local--local-local-78c8b-lh2cc -o yaml | yq '.spec'
podCIDR: 10.3.1.0/24
podCIDRs:
  - 10.3.1.0/24
taints:
  - effect: NoSchedule
    key: node.gardener.cloud/critical-components-not-ready
```

Observe `EveryNodeReady `
```
$ k -n garden-local get shoot local -o yaml | yq '.status.conditions[3]'
lastTransitionTime: "2024-07-30T07:01:55Z"
lastUpdateTime: "2024-07-30T06:56:54Z"
message: too many worker nodes are registered. Exceeding maximum desired machine count (2/1)
reason: NodesScalingDown
status: "False"
type: EveryNodeReady
```

**Anything else we need to know?**:
Expected result:
```
$ k -n garden-local get shoot local -o yaml | yq '.status.conditions[3]'
lastTransitionTime: "2024-07-30T15:01:29Z"
lastUpdateTime: "2024-07-30T15:01:59Z"
message: 1 machine is provisioning and should join the cluster soon
reason: NodesScalingUp
status: Progressing
type: EveryNodeReady
```

**Environment**:

- Gardener version:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- Others:

### PR DESCRIPTION
**Pull request by ScheererJ**
**How to categorize this PR?**

/area dev-productivity
/kind bug

**What this PR does / why we need it**:

Ensure that local CA certificate directory has files before start copying.

In case there are no files in the directory the `cp` operation will fail when trying to `stat` the files.

`gardener-node-agent` fails with:

```
Jul 30 12:38:20 machine-shoot--local--e2e-default-local3-696db-8dpjq gardener-node-agent[349]: {"level":"error","ts":"2024-07-30T12:38:20.167Z","msg":"Reconciler error","controller":"operatingsystemconfig","namespace":"kube-system","name":"gardener-node-agent-local3-b8f97ee2fd9a4297","reconcileID":"67ae7bd7-6150-48d0-a59a-9681ef7846e1","error":"failed executing unit commands: 1 error occurred:\n\t* unable to restart unit \"updatecacerts.service\": restart failed for updatecacerts.service, due failed\n\n","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.17.5/pkg/internal/controller/controller.go:329\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.17.5/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\tsigs.k8s.io/controller-runtime@v0.17.5/pkg/internal/controller/controller.go:227"}
```

`updatecacerts.service` fails with:

```
Aug 01 08:58:47 machine-shoot--local--local-local-79987-jh8pl systemd[1]: Starting updatecacerts.service - Update local certificate authorities...
Aug 01 08:58:47 machine-shoot--local--local-local-79987-jh8pl update-local-ca-certificates.sh[440]: cp: cannot stat '/usr/local/share/ca-certificates/*': No such file or directory
Aug 01 08:58:47 machine-shoot--local--local-local-79987-jh8pl systemd[1]: updatecacerts.service: Main process exited, code=exited, status=1/FAILURE
Aug 01 08:58:47 machine-shoot--local--local-local-79987-jh8pl systemd[1]: updatecacerts.service: Failed with result 'exit-code'.
Aug 01 08:58:47 machine-shoot--local--local-local-79987-jh8pl systemd[1]: Failed to start updatecacerts.service - Update local certificate authorities.
```

**Which issue(s) this PR fixes**:
Fixes https://github.com/gardener/gardener/issues/10171

**Special notes for your reviewer**:

This seems to be more relevant to local setup and end-to-end tests. The result of this issue is that `kubelet` and `containerd` are started and shortly after restarted on any new node. This can lead to "lost" containers, which we saw especially with some critical components in the local setup, e.g. `apiserver-proxy`, where an init container might or might not be started, but there is no follow-up and no logs. `apiserver-proxy` may be especially effected as it has an init container with a very short runtime.

**Release note**:

```bugfix operator
`updatecacerts.service` systemd unit on nodes with Debian OS does not fail anymore if `/usr/local/share/ca-certificates` directory is empty.
```

### CODE REVIEW

**Comment by oliver-goetz** [comment]
Awesome 😄 

/lgtm
/approve

**Comment by Kostov6** [comment]
Doesn't this fix the cause of the failed pod and not the reporting itself? I think you meant to put https://github.com/gardener/gardener/issues/10171 instead of https://github.com/gardener/gardener/issues/10227

**Comment by ialidzhikov** [comment]
> Doesn't this fix the cause of the failed pod and not the reporting itself? I think you meant to put #10171 instead of #10227

I updated the issue to fix/close #10171 instead of #10227.

**Comment by ScheererJ** [comment]
@ialidzhikov / @Kostov6 : From my perspective, this should fix both #10171 and #10227.

From what I saw in end-to-end tests, if we see something similar to #10171 it is due to the fact that one of the nodes does not get ready and hence the old node is not deleted. It does not get ready because the critical components do not get ready. In the cases I looked at, it always seemed to be related to `apiserver-proxy` not getting ready (see #10227).
The root cause as described above seem to be the start+restart of `containerd`/`kubelet`, which caused the init container of `apiserver-proxy` hanging when it was started just before the `containerd`/`kubelet` restart.

This change addresses the reason for the `gardener-node-agent` issue triggering the `containerd`/`kubelet` restart. Another option, as discussed with @rfranzke and @oliver-goetz, would be to add bookkeeping to `gardener-node-agent` so that it does not apply a step unnecessarily a second time if some other step failed. However, this seems to be of lower priority for now.

In summary, I believe this pull request addresses both #10171 and #10227. Feel free to share your reasoning why you think otherwise.
